src/main/java/io/vertx/rabbitmq/impl/RabbitMQueueImpl.java

```diff
@@ -0,0 +1,142 @@
+package io.vertx.rabbitmq.impl;
+
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.rabbitmq.RabbitMQueue;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A implementation of {@link RabbitMQueue}
+ */
+public class RabbitMQueueImpl implements RabbitMQueue {
+
+  private static final Logger log = LoggerFactory.getLogger(RabbitMQueueImpl.class);
+
+  private final int DEFAULT_QUEUE_SIZE = 2048;
+  private Handler<Throwable> exceptionHandler;
+  private Handler<JsonObject> messageArrivedHandler;
+  private Handler<Void> endHandler;
+  private String consumerTag;
+
+  private volatile int queueSize = DEFAULT_QUEUE_SIZE;
+  private AtomicInteger currentQueueSize = new AtomicInteger(0);
+  private AtomicBoolean paused = new AtomicBoolean(false);
+
+  // a storage of all received messages
+  private Queue<JsonObject> messagesQueue = new ConcurrentLinkedQueue<>();
+
+
+  @Override
+  public RabbitMQueue exceptionHandler(Handler<Throwable> exceptionHandler) {
+    this.exceptionHandler = exceptionHandler;
+    return this;
+  }
+
+  @Override
+  public RabbitMQueue handler(Handler<JsonObject> messageArrivedHandler) {
+    this.messageArrivedHandler = messageArrivedHandler;
+    return this;
+  }
+
+  @Override
+  public RabbitMQueue pause() {
+    paused.set(true);
+    return this;
+  }
+
+  @Override
+  public RabbitMQueue resume() {
+    paused.set(false);
+    flushQueue();
+    return this;
+  }
+
+  @Override
+  public RabbitMQueue endHandler(Handler<Void> endHandler) {
+    this.endHandler = endHandler;
+    return this;
+  }
+
+  @Override
+  public synchronized boolean setQueueSize(int size) {
+    queueSize = size;
+    return false;
+  }
+
+  @Override
+  public String consumerTag() {
+    return consumerTag;
+  }
+
+  public void setConsumerTag(String consumerTag) {
+    this.consumerTag = consumerTag;
+  }
+
+  /**
+   * Push message to stream.
+   * <p>
+   * Should be called from a vertx thread.
+   *
+   * @param message received message to deliver
+   */
+  void push(JsonObject message) {
```

Comment:
we should not discard messages when queue is too large, we should simply stop reading from the underlying queue

Reply:
How can we stop reading from underlying? By calling `Channel#basicCancel`right?

Reply:
What do you also think about having an internal queue?